logs/loki/loki.go

```diff
@@ -0,0 +1,60 @@
+package loki
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	netHTTP "net/http"
+	"net/url"
+
+	"github.com/flanksource/commons/http"
+	"github.com/flanksource/duty/context"
+	"github.com/flanksource/duty/types"
+	"github.com/flanksource/incident-commander/logs"
+)
+
+func Fetch(ctx context.Context, baseURL string, auth *types.Authentication, request Request) (*logs.LogResult, error) {
+	parsedBaseURL, err := url.Parse(baseURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse base URL '%s': %w", baseURL, err)
+	}
+
+	apiURL := parsedBaseURL.JoinPath("/loki/api/v1/query_range")
+	apiURL.RawQuery = request.Params().Encode()
+
+	client := http.NewClient()
+	if auth != nil {
+		username, err := ctx.GetEnvValueFromCache(auth.Username, ctx.GetNamespace())
+		if err != nil {
+			return nil, fmt.Errorf("failed to get username: %w", err)
+		}
+		password, err := ctx.GetEnvValueFromCache(auth.Password, ctx.GetNamespace())
+		if err != nil {
+			return nil, fmt.Errorf("failed to get password: %w", err)
+		}
+		client.Auth(username, password)
+	}
+
+	resp, err := client.R(ctx).Get(apiURL.String())
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != netHTTP.StatusOK {
+		bodyBytes, readErr := io.ReadAll(io.LimitReader(resp.Body, 1024*10))
```

Comment:
This should probably handle loki error messages : https://github.com/grafana/loki/blob/main/pkg/ruler/base/api.go#L36